pkg/kubelet/kubelet_server_journal.go

```diff
@@ -0,0 +1,259 @@
+// that log natively vs maybe to a file
+func (a *journalArgs) splitNativeVsFileLoggers(ctx context.Context) ([]string, []string) {
+	var nativeLoggers []string
+	var fileLoggers []string
+
+	for _, service := range a.Services {
+		cmdStr, args := getServiceLogDestinationCmd(service)
+		cmd := exec.CommandContext(ctx, cmdStr, args...)
+
+		output, err := cmd.CombinedOutput()
+		// journalctl will not return an error even if the service is not using journald where as Get-WinEvent will.
+		if err != nil {
+			if _, ok := err.(*exec.ExitError); ok {
+				fileLoggers = append(fileLoggers, service)
+				continue
+			}
+		}
+
+		// Check the journalctl output to figure if the service is using journald or not. This is not needed in the
+		// Get-WinEvent case as the command returns an error if a service is not logging to the Application provider.
+		if checkForNativeLogger(output, service) {
+			nativeLoggers = append(nativeLoggers, service)
+		} else {
+			fileLoggers = append(fileLoggers, service)
+		}
+	}
+	return nativeLoggers, fileLoggers
+}
+
+// copyServiceLogs invokes journalctl or Get-WinEvent with the provided args
+func (a *journalArgs) copyServiceLogs(ctx context.Context, w io.Writer, services []string, previousBoot int64) {
+	cmdStr, args := getLoggingCmd(a, services, previousBoot)
+	cmd := exec.CommandContext(ctx, cmdStr, args...)
+	cmd.Stdout = w
+	cmd.Stderr = w
+
+	if err := cmd.Run(); err != nil {
+		if _, ok := err.(*exec.ExitError); ok {
+			return
+		}
+		if previousBoot == 0 {
+			fmt.Fprintf(w, "error: journal output not available\n")
+		}
+	}
+}
+
+// copyFileLogs loops over all the services and attempts to collect the file logs of each service
+func copyFileLogs(ctx context.Context, w io.Writer, services []string) {
+	if ctx.Err() != nil {
+		return
+	}
+
+	for _, service := range services {
+		if len(service) > 0 {
+			heuristicsCat(ctx, w, service)
+		}
+	}
+}
+
+// heuristicsCat attempts to collect logs from either
+// /var/log/service.log or
+// /var/log/service/service.log or
+// /var/log/service*INFO.* or
+// /var/log/service*INFO or
+// /var/log/service/service*INFO.* or
+// /var/log/service/service*INFO
+// in that order stopping on first success.
+func heuristicsCat(ctx context.Context, w io.Writer, service string) {
+	logDir := "/var/log/"
+	logFiles := [6]string{
+		path.Join(logDir, fmt.Sprintf("%s.log", service)),
+		path.Join(logDir, service, fmt.Sprintf("%s.log", service)),
+		path.Join(logDir, fmt.Sprintf("%s*INFO.*", service)),
+		path.Join(logDir, fmt.Sprintf("%s*INFO", service)),
+		path.Join(logDir, service, fmt.Sprintf("%s*INFO.*", service)),
+		path.Join(logDir, service, fmt.Sprintf("%s*INFO", service)),
+	}
+
+	for _, logFile := range logFiles {
+		if err := cat(ctx, w, logFile); err == nil {
+			break
+		}
+	}
+}
+
+// cat returns the contents of the given logFile
+func cat(ctx context.Context, w io.Writer, logFile string) error {
+	cmdStr, args := getCatCmd(logFile)
```

Comment:
just curios, why execing an external command instead of reading it directly?

Reply:
That is an option too which I can look into as the feature moves from alpha to beta.

Reply:
Re-using `cat` seems definitely like an overkill over reading the file directly, also it'll work on on OS-es ootb.